python/lsst/daf/butler/core/_butlerUri.py

```diff
@@ -0,0 +1,1332 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("ButlerURI",)
+
+import contextlib
+import os
+import os.path
+import shutil
+import urllib
+import pkg_resources
+import posixpath
+from pathlib import Path, PurePath, PurePosixPath
+import requests
+import tempfile
+import copy
+import logging
+
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Iterator,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+)
+
+from .utils import safeMakeDir
+
+if TYPE_CHECKING:
+    try:
+        import boto3
+    except ImportError:
+        pass
+    from .datastore import DatastoreTransaction
+
+
+log = logging.getLogger(__name__)
+
+# Determine if the path separator for the OS looks like POSIX
+IS_POSIX = os.sep == posixpath.sep
```

Comment:
The posixpath docs explicitly say:

> Some of this can actually be useful on non-Posix systems too, e.g. for manipulation of the pathname component of URLs.

so that seems good enough for me since that's exactly what I'm using it for.

Reply:
Yes, true, although `pydoc` also provides a documentation link that 404s. But yes, I suspect this is low-risk.